### PR TITLE
DJ June 2023

### PR DESCRIPTION
**Pull request by xaur**
**Rendered**: https://github.com/xaur/decred-news/blob/draft202306/journal/202306.md

### CODE REVIEW

File: journal/202306.md [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+- Updated the [`AssumeValid`](https://github.com/decred/dcrd/pull/3122) block and [`MinKnownChainWork`](https://github.com/decred/dcrd/pull/3123) as part of the regular release development cycle.
+
+For reference, as of [May 2022](202205.md#development), `AssumeValid` has replaced the legacy "checkpoints" method to optimize initial blockchain downloads. With `AssumeValid`, very old blockchain forks are rejected, and some validation checks are skipped for blocks which are ancestors of the assumed valid block. This can significantly speed up initial blockchain downloads. The updates to `AssumeValid` and `MinKnownChainWork` are externally verified and hardcoded by the devs, and used by default without any action required by node admins. Note: It is possible to change this behavior using the flags `--allowoldforks` and `--assumevalid` as described in `dcrd --help` or [in the code](https://github.com/decred/dcrd/blob/255ef183f66567f1e623026eb524f9cd26fd2b23/chaincfg/mainnetparams.go#L129), but this is not recommended for typical node admins.
+
+{ _what does `MinKnownChainWork` do? can it be disabled like `AssumeValid` and what are the implications?_ }
```
Comment (davecgh) [inline journal/202306.md:91]: It was described a few years back in https://github.com/decred/dcrd/pull/2000.

No, it can't be disabled and there is no reason to since, unlike the old `checkpoints` and rejection of old forks, it can't be invalidated by a chain reorganization.

File: journal/202306.md [hunk 2]
```diff
@@ -0,0 +1,861 @@
...
+
+The ASERT difficulty algorithm is also exponential, and can handle surging activity much better than the existing algorithm. This will greatly disincentivize "hit and run" miners who are only looking for a quick reward at a low difficulty. ASERT also better handles [large drops of hashrate](https://github.com/decred/dcrd/pull/3115#discussion_r1205741807) which cause slow blocks and degraded user experience. The network is stronger when miners are consistently creating new blocks and maintain the target block time of 5 minutes. See also the relevant [discussion on Matrix](https://matrix.to/#/!uCnoeYfbhUnvwWihzS:decred.org/$tbF9LyO1EzQ8m3JDHfX9EYnQz1zr8W0Ra73Yjtprj1M). For those who love fast math and bit tricks, make sure to read the [ASERT code here](https://github.com/decred/dcrd/blob/d9b1921177d4626a22c380853d6c7030a2d5c854/blockchain/standalone/pow.go#L222).
+
+{ _link to DCP-11 if available_ }
+
+{ _do we have the stats how much faster ASERT will adjust to a sharp spike/sharp drop of hashpower, compared to EMA?_ }
```
Comment (davecgh) [inline journal/202306.md:107]: The test data in the PR that implements it has various scenarios.  I'm not sure what number you're after here, but as noted elsewhere, the difficulty adjustment happens every block now instead of every 144, so, the difference can potentially be huge (exponentially so).

File: journal/202306.md [hunk 3]
```diff
@@ -0,0 +1,861 @@
...
+
+{ _what will be the difficulty value immediately after the reset? to make it tangible, what computing power will require to mine, say how many average modern CPUs or GPUs?_ }
```
Comment (davecgh) [inline journal/202306.md:109]: It's in the [chain params](https://github.com/decred/dcrd/blob/a13ca3ace61d28678e1889b9a6303dd88c3dae98/chaincfg/mainnetparams.go#L115). 

Concretely, that is `0x1b00a5a6` which equates to an initial target hash rate of about 1.45 TH/s.  The number of GPUs majorly depends on which GPUs and the speeds that can be obtained by the implementation used to power them.

It doesn't really matter much though because it will very quickly adjust due to the new algorithm and the fact it adjusts every block.

File: journal/202306.md [hunk 4]
```diff
@@ -0,0 +1,861 @@
...
+
+Changes merged in the [dcrtest](https://github.com/decred/dcrtest) framework:
+
+- Updated the voting wallet to support [DCP-12](https://github.com/decred/dcrtest/pull/12) block reward split semantics on simnet.
+- Updated the voting wallet to support [DCP-11](https://github.com/decred/dcrtest/pull/14) proof of work semantics on simnet.
+- Updated to use latest released [Decred module](https://github.com/decred/dcrtest/pull/15) dependencies and latest [main dcrd module](https://github.com/decred/dcrtest/pull/16).
+
+{ _omitted minor PRS: 3139, 3140, 3141, dcrtest#13_ }
+
+{ _image candidate: [Initial Sync Total Bytes Used Comparison](https://matrix.to/#/!zefvTnlxYHPKvJMThI:decred.org/$mbJD22XpiecybtWfRIhFRTawODG9wQujjoKSyOqegI0) - what are these "Total Bytes"? memory?_ }
```
Comment (davecgh) [inline journal/202306.md:138]: It's the total number of bytes allocated in order to perform the initial sync.  Less bytes allocated means less work for the GC and, by extension, less overall work for the CPU which equates to faster sync times.

File: journal/202306.md [hunk 5]
```diff
@@ -0,0 +1,861 @@
...
+- Updated [.gitignore](https://github.com/decred/dcrwallet/pull/2246) to ignore test databases. { _group (internal/dev)_ }
+- Updated the [README](https://github.com/decred/dcrwallet/pull/2247) for people building from source.
+- Bumped the major module version of the in-dev [dcrwallet module](https://github.com/decred/dcrwallet/pull/2248) to v4 to prepare for future changes. { _group (internal/dev)_ }
+- [Updated tests](https://github.com/decred/dcrwallet/pull/2250) to account for Go 1.21 changes. { _may group (tests)_ }
+- Dropped [noisy TSpend logs](https://github.com/decred/dcrwallet/pull/2251) in SPV mode that were used for development. { _group (too small)_ }
+- Removed an [unneeded test](https://github.com/decred/dcrwallet/pull/2253) for the database upgrade code. { _group (tests)_ }
+- Removed a deprecated [`AddressReuse` config](https://github.com/decred/dcrwallet/pull/2258) option from the wallet. { _1. add that it was removed from use earlier but was left in code for compatibility; 2. optional: find an earlier removal/deprecation commit and explain what this config did_ }
+
+Works to consolidate VSP code in the vspd repository:
```
Comment (jholdstock) [inline journal/202306.md:162]: For some colour you could mention that the end goal of these changes is to extract the VSP client out of dcrwallet and into vspd so that it can be reused in dcrdex, enabling dcrdex to purchase tickets and register them with a VSP.

File: journal/202306.md (hunk 2 above)
Comment (davecgh) [reply journal/202306.md:107]: Worst case for that example would be ~75 hours (~3.1 days) under the old algo versus 20-30 mins under the new algo.

File: journal/202306.md [hunk 6]
```diff
@@ -0,0 +1,861 @@
...
+
+### dcrd
+
+_[dcrd](https://github.com/decred/dcrd) is a full node implementation that powers Decred's peer-to-peer network around the world._
+
+First let's look at the changes included in the new 1.8.0 release.
+
+General:
+
+- Updated [Go CI and Go linter](https://github.com/decred/dcrd/commit/ab6d284362de7a30f623289bb1523ecc395c18ca).
```
[the remark this answers is not in the document]
Comment (xaur) [reply journal/202306.md]: Agreed. I planned to reorder it but thanks for the reminder.

File: journal/202306.md [hunk 7]
```diff
@@ -0,0 +1,861 @@
...
+- Updated integration test framework (dcrtest) and [re-enabled treasury test](https://github.com/decred/dcrd/pull/3118). It was [temporarily disabled](https://github.com/decred/dcrd/issues/3093) to get around the circular dependency issue.
+- Miners running v1.8 will [generate new block versions](https://github.com/decred/dcrd/pull/3136) (10), and if the majority of the network upgrades [nodes will reject old block versions](https://github.com/decred/dcrd/pull/3135) (9).
```
[the remark this answers is not in the document]
Comment (xaur) [reply journal/202306.md:82]: Absolutely. Thanks for pointing it out. I reordered the dcrd section as follows:

- Most tangible user-facing changes (upgrade incentives)
- DCP-11 changes bullets
- DCP-11 additional paragraphs with interesting facts about ASERT
- ASERT chart
- Reminder that DCP-12 has been merged earlier + a new link to read the released DCP
- Boring AssumeValid and MinKnownChainWork bumps that we abused to plug some education
- 1.8 changes most relevant to developers
- `master` changes past 1.8
- dcrtest changes closely related to dcrd

I hope it flows much better now but any other feedback is much welcome.

File: journal/202306.md (hunk 1 above)
Comment (xaur) [reply journal/202306.md]: Thanks. I [added](https://github.com/xaur/decred-news/commit/b147200935c98ab6443ec68e43bef8f687e4854e#diff-1d487d8831a0f2f7b2e5f0ea1543054a4e3b287dd31e36d115af3525a053b485R91) an educational paragraph about that.

I used a phrase "minimum known chain work can never go down" and now I'm wondering.

Out of pure theoretical interest, does the "most work" chain always win in Decred given all competing candidates are PoS-approved? Or, can `MinKnownChainWork` get invalidated (i.e. replaced by a smaller value) if PoS voters abandon the "most work" chain and start only approving blocks on a "less work" chain?

File: journal/202306.md (hunk 2 above)
Comment (xaur) [reply journal/202306.md]: I tried looking at test data but could not figure out the practical difference. The improvement is mindblowing. I couldn't blink after reading this and my eyebrow muscles hurt now. So it will adjust in just a few of somewhat slow blocks. I [added](https://github.com/xaur/decred-news/commit/51bf8eb24c0bfba355624051f48891942f383248#diff-1d487d8831a0f2f7b2e5f0ea1543054a4e3b287dd31e36d115af3525a053b485R98) these numbers to the draft, this comparison is very enlightening.

File: journal/202306.md (hunk 3 above)
Comment (xaur) [reply journal/202306.md]: Thanks for sharing this hashrate number. Looks low enough to continue the chain with even very low initial hashrate. I [added](https://github.com/xaur/decred-news/commit/b147200935c98ab6443ec68e43bef8f687e4854e#diff-1d487d8831a0f2f7b2e5f0ea1543054a4e3b287dd31e36d115af3525a053b485R109) both the number and a mention that it is not very important due to how fast ASERT is. But it's still interesting to mention IMO.

File: journal/202306.md (hunk 4 above)
Comment (xaur) [reply journal/202306.md]: Got it, thanks. Will use that in the image caption.

File: journal/202306.md (hunk 5 above)
Comment (xaur) [reply journal/202306.md:169]: Good idea. I [added](https://github.com/xaur/decred-news/commit/b5f72ff) your copy and extended it a bit. Thanks.

File: journal/202306.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (davecgh) [reply journal/202306.md:82]: New flows looks much better.  Nice job.

File: journal/202306.md (hunk 2 above)
Comment (matthawkins90) [reply journal/202306.md]: I think it's going to be super interesting to watch how the [graph of duration between blocks](https://dcrdata.decred.org/charts?chart=duration-btw-blocks&zoom=ikd7pc00-lk4o7400&bin=day&axis=time&visibility=true-false) smooths out over the coming months. Should be really cool.

Comment (davecgh) [reply journal/202306.md]: The overall distribution of the duration between blocks won't change much because mining still follows a Poisson distribution.

The overall expected values based on a perfect Poisson distribution are:

Elapsed | Expected
---------|----------
|10 sec|3.28%|
|30 sec|9.52%|
|1 min|18.13%|
|2 min|32.97%|
|3 min|45.12%|
|4 min|55.07%|
|5 min|63.21%|
|6 min|69.88%|
|7 min|75.34%|
|10 min|86.47%|
|20 min|98.17%|
|30 min|99.75%|

In other words, those are the percentages of blocks that should be found within the given time frames.
